source/opt/common_uniform_elim_pass.cpp

```diff
@@ -130,6 +130,51 @@ ir::Instruction* CommonUniformElimPass::GetPtr(
   return ptrInst;
 }
 
+bool CommonUniformElimPass::IsVolatileStructType(const ir::Instruction& AccessChainInst) {
+  assert(AccessChainInst.opcode() == SpvOpAccessChain);
+  uint32_t memberTypeId = AccessChainInst.GetSingleWordOperand(3);
+  // OpConstant
+  const ir::Instruction* memberTypeDef = def_use_mgr_->GetDef(memberTypeId);
+  uint32_t member = memberTypeDef->GetSingleWordOperand(2);
+  uint32_t VarId = AccessChainInst.GetSingleWordOperand(2);
+  // OpVariable
+  const ir::Instruction* VarInst = def_use_mgr_->GetDef(VarId);
+  uint32_t ptrTypeId = VarInst->GetSingleWordOperand(0);
+  // OpTypePointer
+  const ir::Instruction* TypePtrInst = def_use_mgr_->GetDef(ptrTypeId);
+  uint32_t typeId = TypePtrInst->GetSingleWordInOperand(kTypePointerTypeIdInIdx);
```

Comment:
This sees the volatile when:
 - the base pointer is of struct type S
 - the volatile is on a member of S

It misses cases where the volatile is on a nested struct.  E.g.
 struct Inner {
      volatile float x;
  }; 
  struct Outer {
    Inner a;
  };

layout(set = 0, binding = 0)  buffer B {
  Outer o;
} Bdat;
...
   float f = Bdat.o.a.x;


Either:
 -  add a TODO for the nested struct case, or
 -  handle the nested struct case.   

For the nested struct case, you'll have to walk the type tree, determining the pointer type at each step.   We really should have a common utility for this, but we don't yet.  For now you can look at this sample work-in-progress code that I wrote that does something similar:

https://github.com/dneto0/SPIRV-Tools/blob/robust-access/source/opt/graphics_robust_access_pass.cpp#L188

Reply:
Good point, although I don't know how to write this in glsl. I made a test with a single inner volatile struct and took you approach to walk the type tree.